From 7b8ea624acbfb5aed8c8a52c8311bccb7e87869b Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Thu, 19 Sep 2013 20:12:16 +0530
Subject: [PATCH] fix DateRevQueue to follow FIFO policy in case of tie.

DateRevQueue is expected to give out the commits that have higher commit time.
But in case of tie(same commit time), it should give the commit that is inserted
first. This is inferred from the testInsertTie test case written for
DateRevQueue. Also that test case, right now uses just two commits which caused
it not to fail with the current implementation, so added another commit to
fix it.

Change-Id: Ic71cde0e448c23173a85959d13352366556164e2
---
 .../tst/org/eclipse/jgit/api/RebaseCommandTest.java                 | 6 +++---
 .../tst/org/eclipse/jgit/revwalk/DateRevQueueTest.java              | 5 +++++
 org.eclipse.jgit/src/org/eclipse/jgit/revwalk/DateRevQueue.java     | 2 +-
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/org.eclipse.jgit.test/tst/org/eclipse/jgit/api/RebaseCommandTest.java b/org.eclipse.jgit.test/tst/org/eclipse/jgit/api/RebaseCommandTest.java
index ef5a1ea..59957d6 100644
--- a/org.eclipse.jgit.test/tst/org/eclipse/jgit/api/RebaseCommandTest.java
+++ b/org.eclipse.jgit.test/tst/org/eclipse/jgit/api/RebaseCommandTest.java
@@ -273,8 +273,8 @@ public void testRebaseShouldIgnoreMergeCommits()
 
 		RevWalk rw = new RevWalk(db);
 		rw.markStart(rw.parseCommit(db.resolve("refs/heads/topic")));
-		assertDerivedFrom(rw.next(), e);
 		assertDerivedFrom(rw.next(), d);
+		assertDerivedFrom(rw.next(), e);
 		assertDerivedFrom(rw.next(), c);
 		assertEquals(b, rw.next());
 		assertEquals(a, rw.next());
@@ -289,9 +289,9 @@ public void testRebaseShouldIgnoreMergeCommits()
 				.getReverseEntries();
 		assertEquals("rebase finished: returning to refs/heads/topic", headLog
 				.get(0).getComment());
-		assertEquals("rebase: update file2 on side", headLog.get(1)
+		assertEquals("rebase: Add file2", headLog.get(1).getComment());
+		assertEquals("rebase: update file2 on side", headLog.get(2)
 				.getComment());
-		assertEquals("rebase: Add file2", headLog.get(2).getComment());
 		assertEquals("rebase: update file3 on topic", headLog.get(3)
 				.getComment());
 		assertEquals("checkout: moving from topic to " + b.getName(), headLog
diff --git a/org.eclipse.jgit.test/tst/org/eclipse/jgit/revwalk/DateRevQueueTest.java b/org.eclipse.jgit.test/tst/org/eclipse/jgit/revwalk/DateRevQueueTest.java
index 766a279..72a7401 100644
--- a/org.eclipse.jgit.test/tst/org/eclipse/jgit/revwalk/DateRevQueueTest.java
+++ b/org.eclipse.jgit.test/tst/org/eclipse/jgit/revwalk/DateRevQueueTest.java
@@ -90,20 +90,25 @@ public void testInsertOutOfOrder() throws Exception {
 	public void testInsertTie() throws Exception {
 		final RevCommit a = parseBody(commit());
 		final RevCommit b = parseBody(commit(0, a));
+		final RevCommit c = parseBody(commit(0, b));
 		{
 			q = create();
 			q.add(a);
 			q.add(b);
+			q.add(c);
 
 			assertCommit(a, q.next());
 			assertCommit(b, q.next());
+			assertCommit(c, q.next());
 			assertNull(q.next());
 		}
 		{
 			q = create();
+			q.add(c);
 			q.add(b);
 			q.add(a);
 
+			assertCommit(c, q.next());
 			assertCommit(b, q.next());
 			assertCommit(a, q.next());
 			assertNull(q.next());
diff --git a/org.eclipse.jgit/src/org/eclipse/jgit/revwalk/DateRevQueue.java b/org.eclipse.jgit/src/org/eclipse/jgit/revwalk/DateRevQueue.java
index 0802bfa..6e6afcf 100644
--- a/org.eclipse.jgit/src/org/eclipse/jgit/revwalk/DateRevQueue.java
+++ b/org.eclipse.jgit/src/org/eclipse/jgit/revwalk/DateRevQueue.java
@@ -117,7 +117,7 @@ else if (t > when)
 			head = n;
 		} else {
 			Entry p = q.next;
-			while (p != null && p.commit.commitTime > when) {
+			while (p != null && p.commit.commitTime >= when) {
 				q = p;
 				p = q.next;
 			}
-- 
1.8.2.1

